src/hotspot/share/runtime/os.cpp

```diff
@@ -1764,11 +1765,197 @@ char* os::attempt_reserve_memory_at(char* addr, size_t bytes, bool executable) {
   char* result = pd_attempt_reserve_memory_at(addr, bytes, executable);
   if (result != nullptr) {
     MemTracker::record_virtual_memory_reserve((address)result, bytes, CALLER_PC);
+    log_debug(os)("Reserved memory at " INTPTR_FORMAT " for " SIZE_FORMAT " bytes.", p2i(addr), bytes);
   } else {
     log_debug(os)("Attempt to reserve memory at " INTPTR_FORMAT " for "
                  SIZE_FORMAT " bytes failed, errno %d", p2i(addr), bytes, get_last_error());
   }
+
+  return result;
+}
+
+#ifdef ASSERT
+static void print_points(const char* s, unsigned* points, unsigned num) {
+  stringStream ss;
+  for (unsigned i = 0; i < num; i ++) {
+    ss.print("%u ", points[i]);
+  }
+    log_trace(os, map)("%s, %u Points: %s", s, num, ss.base());
+}
+#endif
+
+// Helper for os::attempt_reserve_memory_between
+// Given an array of things, shuffle them (Fisher-Yates)
+template <typename T>
+static void shuffle_fisher_yates(T* arr, unsigned num, FastRandom& frand) {
+  for (unsigned i = num - 1; i >= 1; i--) {
+    unsigned j = frand.next() % i;
+    swap(arr[i], arr[j]);
+  }
+}
+
+// Helper for os::attempt_reserve_memory_between
+// Given an array of things, do a hemisphere split such that the resulting
+// order is: [first, last, first + 1, last - 1, ...]
+template <typename T>
+static void hemi_split(T* arr, unsigned num) {
+  T* tmp = (T*)::alloca(sizeof(T) * num);
+  for (unsigned i = 0; i < num; i++) {
+    tmp[i] = arr[i];
+  }
+  for (unsigned i = 0; i < num; i++) {
+    arr[i] = is_even(i) ? tmp[i / 2] : tmp[num - (i / 2) - 1];
+  }
+}
+
+// Given an address range [min, max), attempts to reserve memory within this area, with the given alignmend.
+// If randomize is true, the location will be randomized.
+char* os::attempt_reserve_memory_between(char* min, char* max, size_t bytes, size_t alignment, bool randomize) {
```

Comment:
`os::attempt_reserve_memory_between` is a generic API, usable for other things too. One obvious use case would be allocating the heap (replacing all that lets-reserve-the-heap-in-lower-regions); another one Shenandoah CollectionSet; and some more. I don't want to impose the "randomize" decision on all callers.

Reply:
> Could be broken out into helper functions (one for randomized, one for non-randomized and one that dispatches between the two)?

See my answer to @iklam https://github.com/openjdk/jdk/pull/15041/files#r1290988328 . I spent considerable time trying to come up with a good split along the lines of randomized/non-randomized. All variants I came up with would lead to code duplication that I'm sure would be criticised by other reviewers. Let's see what others think.